core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java

```diff
@@ -11,20 +11,144 @@
 
 package alluxio.worker.block;
 
+import alluxio.StorageTierAssoc;
+import alluxio.WorkerStorageTierAssoc;
+import alluxio.client.file.FileSystemContext;
+import alluxio.exception.AlluxioException;
+import alluxio.exception.BlockAlreadyExistsException;
+import alluxio.exception.BlockDoesNotExistException;
 import alluxio.proto.dataserver.Protocol;
+import alluxio.util.IdUtils;
+import alluxio.util.io.BufferUtils;
+import alluxio.wire.WorkerNetAddress;
+import alluxio.worker.block.io.BlockReader;
+import alluxio.worker.block.io.BlockWriter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
 
 /**
  * Handles client requests to asynchronously cache blocks. Responsible for managing the local
  * worker resources and intelligent pruning of duplicate or meaningless requests.
  */
+@ThreadSafe
 public class AsyncCacheRequestManager {
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncCacheRequestManager.class);
+
+  private final StorageTierAssoc mStorageTierAssoc = new WorkerStorageTierAssoc();
+  /** Executor service for execute the async cache tasks. */
+  private final ExecutorService mAsyncCacheExecutor;
+  /** The block worker. */
+  private final BlockWorker mBlockWorker;
+  private final Object mLock = new Object();
+  @GuardedBy("mLock")
+  private final Set<Long> mPendingRequests;
+  private final WorkerNetAddress mLocalWorkerAddress;
+
+  public AsyncCacheRequestManager(ExecutorService service, BlockWorker blockWorker) {
+    mAsyncCacheExecutor = service;
+    mBlockWorker = blockWorker;
+    mPendingRequests = new HashSet<>();
+    try {
+      mLocalWorkerAddress = FileSystemContext.INSTANCE.getLocalWorker();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
   /**
    * Handles a request to cache a block asynchronously. This is a non-blocking call.
    *
    * @param request the async cache request
    * fields will be available
    */
   public void submitRequest(Protocol.AsyncCacheRequest request) {
+    long blockId = request.getBlockId();
+    synchronized (mLock) {
+      if (mPendingRequests.contains(blockId)) {
+        // This block is already planned.
+        return;
+      }
+      mPendingRequests.add(request.getBlockId());
+    }
+    mAsyncCacheExecutor.submit(() -> {
+      Protocol.OpenUfsBlockOptions openUfsBlockOptions = request.getOpenUfsBlockOptions();
+      long sessionId = IdUtils.createSessionId();
```

Comment:
Use a negative id to indicate internal service usage?

Reply:
done